crates/oxc_linter/src/rules/eslint/dot_notation.rs

```diff
@@ -0,0 +1,343 @@
+}
+
+#[derive(Debug, Default, Clone)]
+pub struct DotNotation(Box<DotNotationConfig>);
+
+#[derive(Debug, Clone)]
+pub struct DotNotationConfig {
+    allow_keywords: bool,
+    allow_pattern: Option<Regex>,
+}
+
+impl Default for DotNotationConfig {
+    fn default() -> Self {
+        Self { allow_keywords: true, allow_pattern: None }
+    }
+}
+
+declare_oxc_lint!(
+    /// ### What it does
+    /// This rule enforces dot notation whenever possible.
+    ///
+    /// ### Why is this bad?
+    /// In JavaScript, one can access properties using the dot notation (foo.bar) or square-bracket
+    /// notation (foo["bar"]). However, the dot notation is often preferred because it is easier to
+    /// read, less verbose, and works better with aggressive JavaScript minimizers.
+    ///
+    /// ### Examples
+    ///
+    /// Examples of **incorrect** code for this rule:
+    /// ```js
+    /// const x = foo["bar"];
+    ///
+    /// // ** For sample config { "allowKeywords": false } **
+    /// const foo = { "class": "CS 101" }
+    /// const x = foo["class"]; // Property name is a reserved word, square-bracket notation
+    ///                         // required
+    ///
+    /// class C {
+    ///     #in;
+    ///     foo() {
+    ///         this.#in; // Dot notation is required for private identifiers
+    ///     }
+    /// }
+    ///
+    /// // ** For sample config { "allowPattern": "^[a-z]+(_[a-z]+)+$" } **
+    /// const data = {};
+    /// data["fooBar"] = 42;
+    /// ```
+    ///
+    /// Examples of **correct** code for this rule:
+    /// ```js
+    /// const x = foo.bar;
+    /// const y = foo[bar];
+    ///
+    /// // For sample config { "allowPattern": "^[a-z]+(_[a-z]+)+$" }
+    /// const data = {};
+    /// data["foo_bar"] = 42;
+    /// ```
+    DotNotation,
+    eslint,
+    style,
+    // Fixes are possible
+    // See <https://github.com/eslint/eslint/blob/main/lib/rules/dot-notation.js>
+    pending
+);
+
+/// List of ES3 keywords.
+const KEYWORDS: &[&str] = &[
+    "abstract",
+    "boolean",
+    "break",
+    "byte",
+    "case",
+    "catch",
+    "char",
+    "class",
+    "const",
+    "continue",
+    "debugger",
+    "default",
+    "delete",
+    "do",
+    "double",
+    "else",
+    "enum",
+    "export",
+    "extends",
+    "false",
+    "final",
+    "finally",
+    "float",
+    "for",
+    "function",
+    "goto",
+    "if",
+    "implements",
+    "import",
+    "in",
+    "instanceof",
+    "int",
+    "interface",
+    "long",
+    "native",
+    "new",
+    "null",
+    "package",
+    "private",
+    "protected",
+    "public",
+    "return",
+    "short",
+    "static",
+    "super",
+    "switch",
+    "synchronized",
+    "this",
+    "throw",
+    "throws",
+    "transient",
+    "true",
+    "try",
+    "typeof",
+    "var",
+    "void",
+    "volatile",
+    "while",
+    "with",
+];
+lazy_static! {
+    static ref VALID_IDENTIFER: Regex = Regex::new(r"^[a-zA-Z_$][a-zA-Z0-9_$]*$").unwrap();
+}
```

Comment:
We have a strong preference to avoid using `Regex` unless it's necessary (for example, the `allowPattern` needs to support that syntax), as it is relatively slow (https://oxc.rs/docs/contribute/rules.html#performance). We can rewrite this using a mini-parser instead, here is my suggestion:

```suggestion
fn is_valid_identifier(s: &str) -> bool {
    s.as_bytes().first().map_or(false, |&c| (c.is_ascii_alphabetic() || c == b'_' || c == b'$'))
        && s.bytes().skip(1).all(|c| c.is_ascii_alphanumeric() || c == b'_' || c == b'$')
}
```